rshim.c

```diff
@@ -2765,6 +2610,141 @@ static const struct cuse_methods rshim_misc_fops = {
 };
 #endif
 
+/* Rshim file operations routines */
+
+/* ioctl message header. */
+typedef struct {
+  uint32_t addr;
+  uint64_t data;
+} __attribute__((packed)) rshim_ioctl_msg;
+
+enum {
+  RSHIM_IOC_READ = _IOWR('R', 0, rshim_ioctl_msg),
+  RSHIM_IOC_WRITE = _IOWR('R', 1, rshim_ioctl_msg),
+};
+
+#ifdef HAVE_RSHIM_FUSE
+static void rshim_rshim_ioctl(fuse_req_t req, int cmd, void *arg,
+                              struct fuse_file_info *fi,
+                              unsigned flags, const void *in_buf,
+                              size_t in_bufsz, size_t out_bufsz)
+{
+  struct rshim_backend *bd = fuse_req_userdata(req);
+  rshim_ioctl_msg msg;
+  struct iovec iov;
+  int rc = 0;
+
+  switch (cmd) {
+  case RSHIM_IOC_READ:
+  case RSHIM_IOC_WRITE:
+    iov.iov_base = arg;
+    iov.iov_len = sizeof(msg);
+
+    if (!in_bufsz) {
+      fuse_reply_ioctl_retry(req, &iov, 1, NULL, 0);
+      return;
+    }
+
+    if (in_bufsz != sizeof(msg)) {
+      fuse_reply_err(req, EINVAL);
+      return;
+    }
+
+    memcpy(&msg, in_buf, sizeof(msg));
+
+    if (cmd == RSHIM_IOC_WRITE) {
+      pthread_mutex_lock(&bd->mutex);
+      rc = bd->write_rshim(bd,
+                           (msg.addr >> 16) & 0xF, /* channel # */
+                           msg.addr & 0xFFFF, /* addr */
+                           msg.data);
+      pthread_mutex_unlock(&bd->mutex);
+    } else {
+      if (!out_bufsz) {
+        fuse_reply_ioctl_retry(req, &iov, 1, &iov, 1);
+        return;
+      }
+
+      pthread_mutex_lock(&bd->mutex);
+      rc = bd->read_rshim(bd,
+                           (msg.addr >> 16) & 0xF, /* channel # */
+                           msg.addr & 0xFFFF, /* addr */
+                           &msg.data);
+      pthread_mutex_unlock(&bd->mutex);
+    }
+
+    if (!rc)
+      fuse_reply_ioctl(req, 0, &msg, sizeof(msg));
+    else
+      fuse_reply_err(req, rc);
+    break;
+
+  default:
+    fuse_reply_err(req, ENOSYS);
+    break;
+  }
+}
+#endif
+#ifdef HAVE_RSHIM_CUSE
+static void rshim_rshim_ioctl(struct cuse_dev *cdev, int fflags,
+                              unsigned long cmd, void *peer_data)
+{
+  struct rshim_backend *bd = cuse_dev_get_priv0(cdev);
+  int rc, retval = CUSE_ERR_INVALID;
+  rshim_ioctl_msg msg;
+
+  pthread_mutex_lock(&bd->mutex);
+
+  switch (cmd) {
+  case RSHIM_IOC_READ:
+    retval = cuse_copy_in(peer_data, &msg, sizeof(msg));
+    if (retval == CUSE_ERR_NONE) {
+      rc = bd->read_rshim(bd,
+                           (msg.addr >> 16) & 0xF, /* channel # */
+                           msg.addr & 0xFFFF, /* addr */
+                           &msg.data);
+      if (!rc)
+        retval = cuse_copy_out(&msg, peer_data, sizeof(msg));
+      else
```

Comment:
Same here, missing ";" at end of line below:

Reply:
done